src/storage/write_ahead_log/log_serializer_task.cpp

```diff
@@ -65,24 +68,31 @@ void LogSerializerTask::HandFilledBufferToWriter() {
 void LogSerializerTask::SerializeBuffer(IterableBufferSegment<LogRecord> *buffer_to_serialize) {
   // Iterate over all redo records in the redo buffer through the provided iterator
   for (LogRecord &record : *buffer_to_serialize) {
-    if (record.RecordType() == LogRecordType::COMMIT) {
-      auto *commit_record = record.GetUnderlyingRecordBodyAs<CommitRecord>();
-
-      // If a transaction is read-only, then the only record it generates is its commit record. This commit record is
-      // necessary for the transaction's callback function to be invoked, but there is no need to serialize it, as
-      // it corresponds to a transaction with nothing to redo.
-      if (!commit_record->IsReadOnly()) SerializeRecord(record);
-      commits_in_buffer_.emplace_back(commit_record->Callback(), commit_record->CallbackArg());
-      // Not safe to mark read only transactions as the transactions are deallocated preemptively without waiting for
-      // logging (there is nothing to log after all)
-      if (!commit_record->IsReadOnly()) commit_record->Txn()->log_processed_ = true;
-    } else if (record.RecordType() == LogRecordType::ABORT) {
-      // If an abort record shows up at all, the transaction cannot be read-only
-      SerializeRecord(record);
-      record.GetUnderlyingRecordBodyAs<AbortRecord>()->Txn()->log_processed_ = true;
-    } else {
-      // Any record that is not a commit record is always serialized.`
-      SerializeRecord(record);
+    switch (record.RecordType()) {
+      case (LogRecordType::COMMIT): {
+        auto *commit_record = record.GetUnderlyingRecordBodyAs<CommitRecord>();
+
+        // If a transaction is read-only, then the only record it generates is its commit record. This commit record is
+        // necessary for the transaction's callback function to be invoked, but there is no need to serialize it, as
+        // it corresponds to a transaction with nothing to redo.
+        if (!commit_record->IsReadOnly()) SerializeRecord(record);
+        commits_in_buffer_.emplace_back(commit_record->CommitCallback(), commit_record->CommitCallbackArg());
+        // Once serialization is done, we notify the txn manager to let GC know this txn is ready to clean up
+        commit_record->TxnManager()->NotifyTransactionSerialized(commit_record->Txn());
```

Comment:
Also mentioned the transaction manager code, but the log serializer should be talking directly to the timestamp manager to make the dependency explicit.

Reply:
Done.